### PR TITLE
🚧 feat: add vapp properties

### PR DESCRIPTION
**Pull request by MrKeiKun**
This pull request introduces support for vApp properties in the vSphere ISO builder. The changes include updates to the configuration structures, documentation, and relevant functions to handle vApp properties.

Support for vApp Properties:

* Added `vAppConfig` struct to define vApp properties for virtual machines. (`builder/vsphere/iso/step_create.go`) [[1]](diffhunk://#diff-d5153ce3b0321af12537b31ee71ba82155dc1a8ee7ae028dfe9071dec0aa2fc9R22-R32) [[2]](diffhunk://#diff-d5153ce3b0321af12537b31ee71ba82155dc1a8ee7ae028dfe9071dec0aa2fc9R113-R114)
* Updated `CreateConfig` and `FlatCreateConfig` to include `VAppConfig` for setting vApp properties. (`builder/vsphere/driver/vm.go`) [[1]](diffhunk://#diff-f81447593c9b389c1718a91e14307807b280b36f7c76f64a150637d6faba312bR136) (`builder/vsphere/iso/config.hcl2spec.go`) [[2]](diffhunk://#diff-2fd27dfc0b7ffb82b2fce57af1eb5d95c320069e1625bef83964d01c76b57d63R44) (`builder/vsphere/iso/step_create.hcl2spec.go`) [[3]](diffhunk://#diff-9c410a3d0f75d5d0aff60096ee4a43737c4a3e896167785848f4d72250f89c93R22)
* Modified `HCL2Spec` functions to handle the new `vAppConfig` block. (`builder/vsphere/iso/config.hcl2spec.go`) [[1]](diffhunk://#diff-2fd27dfc0b7ffb82b2fce57af1eb5d95c320069e1625bef83964d01c76b57d63R197) (`builder/vsphere/iso/step_create.hcl2spec.go`) [[2]](diffhunk://#diff-9c410a3d0f75d5d0aff60096ee4a43737c4a3e896167785848f4d72250f89c93R45) [[3]](diffhunk://#diff-9c410a3d0f75d5d0aff60096ee4a43737c4a3e896167785848f4d72250f89c93R78-R100)
* Updated the `Run` method in `StepCreateVM` to pass `VAppProperties` to the `CreateVM` function. (`builder/vsphere/iso/step_create.go`)
* Added generated documentation for `vAppConfig` and updated existing documentation to include the new configuration option. (`.web-docs/components/builder/vsphere-iso/README.md`) [[1]](diffhunk://#diff-6752170ad5617582d4773e4abc3c9a943a0685112b0b5591bd80239ee6aa08deR835-R836) (`docs-partials/builder/vsphere/iso/CreateConfig-not-required.mdx`) [[2]](diffhunk://#diff-7b3e3cdc97cca456bbd4308cb3151abd36adba9d85a95752cbfdcffbf2d437bcR34-R35) (`docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx`) [[3]](diffhunk://#diff-ee5d72a4e2069d0853cb117c52df137480f4d6e8b32a486eea095ae513e20b4bR1-R11)

### CODE REVIEW

**Comment by harlequin** [comment]
Please merge, feature would be useful

**Comment by tenthirtyam** [comment]
@MrKeiKun Couls to please provide A summary and test results in the pull request description?

**Comment by tenthirtyam** [comment]
CI is failing.

File: builder/vsphere/iso/step_create.go
Comment (lbajolet-hashicorp) [inline builder/vsphere/iso/step_create.go:21]: Is there a reason why `vAppConfig` is introduced here? If we need the properties only, we could introduce a `vapp_properties` attribute to the main config directly rather than as a separate attribute.
Also I may be mistaken by I believe `packer-sdc` only considers public types for generation, I wonder how the flat configs were generated in the first place? But this would explain the `make generate` error we see in the CI as it removes the structure from the generated code.

**Review by lbajolet-hashicorp — CHANGES_REQUESTED** [review]
Hi @MrKeiKun,

The initial draft looks OK to me, I left a fix suggestion for the generation problem, I believe that if you change the type to be a public structure, it should mend the generated code.

That being said, I'm not sure we need to introduce an extra structure to the package, it seems the only attribute could be directly part of the general config, especially since we don't define any function on the structure itself.
On another note, I am very much unfamiliar with VSphere so please bear with me here, but are there other kind of configs we want to support regarding the VAppConfig? I'm also a bit cautious regarding the `properties` type, since it's a map of string->string, this implies that all attributes' value must be a string, and from the screenshot I saw, there seems to be multiple possibilities, is that right? I see on the issue someone trying to input some XML, is there a standard format to specify that, with type information? If so we should probably aim to support this also (though this can be for later I presume).

File: builder/vsphere/iso/step_create.go
Comment (tenthirtyam) [reply builder/vsphere/iso/step_create.go:21]: This is how it's implemented `vsphere-clone` builder.

I just pushed a change that should hopefully resolve the CI. 🤞🏻

**Comment by tenthirtyam** [comment]
✅ CI is now **Successful**.
✅ Marking as **Ready for Review**.

File: builder/vsphere/iso/step_create.go
Comment (tenthirtyam) [reply builder/vsphere/iso/step_create.go:21]: https://github.com/hashicorp/packer-plugin-vsphere/pull/299#issuecomment-1802409752

File: docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx
Comment (nywilken) [inline docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx:3]: Regarding Lucas comment about the properties field being a map[string]string I will defer to your expertise on vSphere and assume that it is not an issue. But please let me know if it is something that we need to call out in the docs further. 

Also does "These generally come from an existing template that was created from an imported OVF or OVA file." apply to the ISO builder?

Comment (tenthirtyam) [reply docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx:3]: i'll need to check, but I don't think it directly applies. I also agree that moving it to common would mean more ideal.

I'm happy to hold this one for further evaluation and testing to be honest. I only addressed the failing CI.

**Comment by tenthirtyam** [comment]
Marking this as **draft** for further review and outcome testing before merging.

File: docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx
Comment (tenthirtyam) [reply docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx:3]: I started to take a look at the PR this evening and the descriptions need some updates since it's a copy/paste from the comments in the clone builder for the clone step. I'll make these changes accordingly.

I need to chat with @dilyar85 to see if this should be in common, but I don't think it will apply to the supervisor builder and therefore not be in common.

**Comment by harlequin** [comment]
Any Updates to this?

**Comment by adeturner** [comment]
Agree, I've still got the issue in #305. Would be good to get this merged even in a "beta state".  Currently with OVF templates the only work around is to use ssh/remote-exec provisioners which are marked by hashicorp "last resort."

**Comment by harlequin** [comment]
@adeturner Can you please provide some examples into this issue?
I guess this "last resort" solution will help many people until it's implemented ...

**Comment by adeturner** [comment]
@harlequin no worries, see [remote_exec_workaround](https://gist.github.com/adeturner/844ec33d4210a1b6741842980fc30efd) and [the_vapp_ideal](https://gist.github.com/adeturner/2e57dd7d52793d2fc79d7481b8d24013)

File: docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx
Comment (dilyar85) [reply docs-partials/builder/vsphere/iso/vAppConfig-not-required.mdx:3]: This doesn't apply to the current supervisor builder. To set a VM's vApp properties in Supervisor, users need to specify the `bootstrap_provider` as "vAppConfig" and provide the properties in the `bootstrap_data_file`. This is required by the [VM-Operator v1alpha1 API](https://vm-operator.readthedocs.io/en/latest/concepts/workloads/guest/#vappconfig) that the supervisor builder uses internally.

In VM-Operator v1alpha2 (in the upcoming vSphere 8 Update 3 release), there is a new `properties` field introduced in the [VirtualMachineBootstrapVAppConfigSpec](https://vm-operator.readthedocs.io/en/latest/ref/api/v1alpha2/#virtualmachinebootstrapvappconfigspec). So making this common could be helpful as the supervisor builder would be able to set that field from this common config. But we would likely introduce a new config when we switch the supervisor builder to v1a2, since that field also supports setting value from K8s secret key selector.

I am fine with leaving this as is and making it common if needed when we add v1a2 support in the supervisor builder.

**Comment by tenthirtyam** [comment]
Thus far, I've not been able to perform a build that will apply these properties to an image with the `vsphere-iso` builder after building from the source/branch.

@MrKeiKun 👋 - could you please provide a summary, minimal configuration example, test results in the pull request description?

Without this information I am reticent to move this one forward until the this can be successfully tested.

**Comment by tenthirtyam** [comment]
After taking a look at this pull request it seems to be insufficient since you must also enable the vApp Options before being able to apply the properties.

Based on this, I'll look at addressing this item for the v1.5.0 release whilst [taking into consideration Lucas' comments](https://github.com/hashicorp/packer-plugin-vsphere/pull/299#pullrequestreview-1688433128). 

**Ryan Johnson**
Distinguished Engineer, VMware by Broadcom

**Pull request closed without merging**
